### PR TITLE
Katie's Hotel

### PR DESCRIPTION
**Pull request by katietj**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | When I started the project, I just had a room class and reservation class. When I first began I thought I might need a hotel manager/front desk class, but since I didn't have anything to go there when the project began, I held off on deciding until I created some code for the first two classes. Once they got too big and I was thinking of methods that couldn't fit in either, I made a new class called front_desk.
Describe a concept that you gained more clarity on as you worked on this assignment. | It became more clear to me how classes worked together, this is something that confused me a lot during oo-ride share. I also became much better at debugging errors that come up when creating tests.
Describe a _nominal test_ that you wrote for this assignment. | A nominal test I wrote is where I created an instance of room and tested to be sure that it was returning an instance of room. I did this type of nominal test for everything I instantiated as a first test
Describe an _edge case test_ that you wrote for this assignment. | I created a test where I tried to book a room with an invalid room number to make sure that an error would be raised when this happened. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Pseudocode is so helpful! I used it for most of my code and when I did, it was smoother and more focused work time. I found that writing a small amount of a test, then a small amount of code, then finish the test and finish the code worked best for me. I had a day where I wrote a lot of code before testing and that was a huge mistake I won't make again. If you have a ton of code with no tests, its hard to find where the problems are.

### CODE REVIEW

File: lib/block.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+require 'time'
+require_relative 'reservation'
+require_relative 'front_desk'
+require_relative 'room'
+
+class Block
+  attr_reader :block_start, :block_end, :rate, :block_rooms, :block_ID
```
Comment (CheezItMan) [inline lib/block.rb:7]: Minor note, you have a reader and writer for `block_ID`, instead maybe just make an accessor.

File: lib/block.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  def block_id_generator
+    letters = ('A'..'Z').to_a.shuffle[0..1].join
+    id = letters + (0..9).to_a.shuffle[0..2].join
+    return id
+  end
+
+  def is_available_block?(room)
+    return false if @reserved_in_block.include?(room)
```
Comment (CheezItMan) [inline lib/block.rb:30]: You need an else here.  Otherwise it'll return `nil`

File: lib/block.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    @block_rooms = converted_rooms
+    #when i actually reserve a room, it goes into reserved_in_block array? this is an option for
+    # refactor.
+    @reserved_in_block = []
+    if block_rooms.length > 5
+      raise StandardError.new "Block is limited to 5 rooms"
+    end
+  end
+
+  def block_id_generator
```
Comment (CheezItMan) [inline lib/block.rb:23]: Just want to point out, this won't be a necessarily unique id.

File: lib/front_desk.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  def load_rooms
+    rooms = []
+    20.times do |i|
+      rooms << Room.new((i + 1))
+    end
+    return rooms
+  end
+
+  def list_all_rooms
+    rooms.each do |room|
```
Comment (CheezItMan) [inline lib/front_desk.rb]: This loop isn't doing anything.  you could just `return rooms`

File: lib/front_desk.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    room = find_room(room_number)
+    room.add_reservation_to_room(new_reservation)
+
+  end
+
+  def find_room(room_number)
+    return @rooms.find {|room| room.room_number == room_number}
+  end
+
+  def find_by_date(date)
```
Comment (CheezItMan) [inline lib/front_desk.rb]: This method only finds reservations that **start** on the given date.  Instead you should return reservations are occur **during** that date.

File: lib/front_desk.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  end
+
+  def check_availability(room_number, start_day, end_day)
+    available_rooms = find_available_rooms(Date.parse(start_day), Date.parse(end_day))
+    raise StandardError, 'Room not available' unless available_rooms.include?(room_number)
+  end
+
+
+  def create_block(block_start, block_end, rate, block_rooms)
+    find_available_rooms(block_start, block_end)
```
Comment (CheezItMan) [inline lib/front_desk.rb:78]: Shouldn't you do something with the return value of `find_available_rooms `

File: lib/front_desk.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  end
+
+
+  def find_available_rooms_in_block(block_ID)
+    available_rooms = []
+    block = find_block_by_id(block_ID)
+    rooms = block.block_rooms
+    rooms.each do |room|
+      room_is_available = false
+      if !block.is_available_block?(room)
```
Comment (CheezItMan) [inline lib/front_desk.rb:112]: I'm not sure this is right.  If the room in the block is not available you're setting `room_is_available` to `true`.

File: lib/reservation.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    def initialize(room, start_day, end_day,nightly_cost)
+      @room = room
+      @nightly_cost = nightly_cost
+      @start_day = Date.parse(start_day)
+      @end_day = Date.parse(end_day)
+      dates_validation(@start_day, @end_day)
+      @total_nights = nights_in_reservation
+    end
+
+    def dates_validation(start_day, end_day)
```
Comment (CheezItMan) [inline lib/reservation.rb:18]: I would make this method `private`

File: lib/room.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
+require 'time'
+require_relative 'reservation'
+require_relative 'front_desk'
+
+
+  class Room
+
+    attr_reader :room_number, :room_reservations
```
Comment (CheezItMan) [inline lib/room.rb]: So rooms have reservations, but not blocks.  I would either pull reservations into your `FrontDesk` class or put blocks into room as well.

File: spec/front_desk_spec.rb [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+    describe 'find reservations by date' do
+      before do
+        @front_desk = FrontDesk.new
+        @front_desk.create_reservation(3, "2012-4-12", "2012-4-22",200)
+        @front_desk.create_reservation(4, "2012-4-12", "2012-4-22",200)
+        @front_desk.create_reservation(5, "2012-4-23", "2012-4-29",200)
+        @front_desk.create_reservation(6, "2012-4-25", "2012-4-28",200)
+      end
+      it "returns reservations for a specific date date" do
```
Comment (CheezItMan) [inline spec/front_desk_spec.rb:69]: Also test if there are no reservations on the date.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Good number of commit and good commit messages
Answer comprehension questions	| Check
**Design** |
Each class is responsible for a single piece of the program | For the most part, although room is also tracking reservations. 
Classes are loosely coupled | Room & reservations are coupled together. 
**Wave 1** |
List rooms | Check
Reserve a room for a given date range | Check
List reservations for a given date | Check
Calculate reservation price | Check
Invalid date range produces an error | Check
**Wave 2** |
View available rooms for a given date range | Check 
Reserving a room that is not available produces an error |  Check
**Wave 3** |
Create a block of rooms |  Check
Check if a block has rooms |  Check
Reserve a room from a block |  Check
Test coverage | 100%
**Fundamentals** | 
Names variables, classes and modules appropriately |  Check
Understanding of variable scope - local vs instance |  Check
Can create complex logical structures utilizing variables |  Check
Appropriately uses methods to break down tasks into smaller simpler tasks |  Check
Understands the differences between class and instance methods |  Check
Appropriately uses iterators and enumerables |  Check
Appropriately writes and utilizes classes |  Check
Appropriately utilizes modules as a mechanism to encapsulate functionality | No module used
**Wrap Up** |
There is a refactors.txt file |  Check
The file provides a roadmap to future changes |  Check
**Additional Feedback** | Nicely done.  See my inline notes, but you hit the learning goals and have a fairly well designed project.
